### PR TITLE
Add settings dialog which allows user to select theme and view licenses

### PR DESCRIPTION
**Pull request by JolandaVerhoef**

### CODE REVIEW

**Comment by dturner** [comment]
Spotless is failing, and confusingly it no longer outputs the command to apply the fixes. You'll need to run: 

```
./gradlew --init-script gradle/init.gradle.kts --no-configuration-cache spotlessApply
```

File: core/data/src/test/java/com/google/samples/apps/nowinandroid/core/data/repository/OfflineFirstUserDataRepositoryTest.kt [hunk 1]
```diff
@@ -129,4 +147,44 @@ class OfflineFirstUserDataRepositoryTest {
                     .first()
             )
         }
+
+    @Test
+    fun offlineFirstUserDataRepository_set_theme_brand_delegates_to_nia_preferences() =
```
Comment (JoseAlcerreca) [inline core/data/src/test/java/com/google/samples/apps/nowinandroid/core/data/repository/OfflineFirstUserDataRepositoryTest.kt:152]: weird test naming, should have max two underscores

File: core/datastore/src/main/java/com/google/samples/apps/nowinandroid/core/datastore/NiaPreferencesDataSource.kt [hunk 2]
```diff
@@ -34,6 +36,21 @@ class NiaPreferencesDataSource @Inject constructor(
                 bookmarkedNewsResources = it.bookmarkedNewsResourceIdsMap.keys,
                 followedTopics = it.followedTopicIdsMap.keys,
                 followedAuthors = it.followedAuthorIdsMap.keys,
+                themeBrand = when (it.themeBrand!!) {
```
Comment (JoseAlcerreca) [inline core/datastore/src/main/java/com/google/samples/apps/nowinandroid/core/datastore/NiaPreferencesDataSource.kt:39]: it.themeBrand == true is safer because if for some reason it's null, it will default to false. And below.

File: core/designsystem/src/main/java/com/google/samples/apps/nowinandroid/core/designsystem/theme/Theme.kt [hunk 3]
```diff
@@ -173,55 +174,63 @@ val DarkAndroidBackgroundTheme = BackgroundTheme(color = Color.Black)
...
+    content: @Composable () -> Unit
+) = NiaTheme(
+    darkTheme = darkTheme,
+    androidTheme = androidTheme,
+    disableDynamicTheming = false,
+    content = content
+)
+
+/**
+ * Now in Android theme. This is an internal only version, to allow disabling dynamic theming
```
Comment (JoseAlcerreca) [inline core/designsystem/src/main/java/com/google/samples/apps/nowinandroid/core/designsystem/theme/Theme.kt:194]: It should live in the test srcSet then

File: app/src/main/java/com/google/samples/apps/nowinandroid/MainActivity.kt [hunk 4]
```diff
@@ -38,16 +58,60 @@ class MainActivity : ComponentActivity() {
...
+                    .onEach {
+                        userDataResult = it
+                    }
+                    .collect()
+            }
+        }
+
+        // Keep the splash screen on-screen until the user data is loaded
+        splashScreen.setKeepOnScreenCondition {
+            when (userDataResult) {
```
[the remark this answers is not in the document]
Comment (JoseAlcerreca) [reply app/src/main/java/com/google/samples/apps/nowinandroid/MainActivity.kt:87]: Edit: Looked into it and turns out
"The condition is evaluated before each request to draw the application, so it needs to be fast to avoid blocking the UI."
This should be documented here, add comment?

File: app/src/main/java/com/google/samples/apps/nowinandroid/MainActivity.kt [hunk 5]
```diff
@@ -38,16 +58,60 @@ class MainActivity : ComponentActivity() {
...
+        val splashScreen = installSplashScreen()
         super.onCreate(savedInstanceState)
 
+        /**
+         * The current user data, updated here to drive the UI theme
+         */
+        var userDataResult: Result<UserData> by mutableStateOf(Result.Loading)
+
+        // Update the user data
+        lifecycleScope.launch {
```
Comment (JoseAlcerreca) [inline app/src/main/java/com/google/samples/apps/nowinandroid/MainActivity.kt:74]: I'd move any UI logic to a VM, even if it's super simple. Just collect the userDataResult from it here.

**Comment by dturner** [comment]
I've done a fairly major refactor of this PR: 

- Created a `:feature:settings` module which contains everything associated with showing settings to the user
- Refactored your existing code out of the `ForYou` module and into this new module
- Updated the `SettingsDialog` to include links to the privacy policy and open source licenses (for legal)
- Added e2e, UI and unit tests

Comments most welcome.

File: app/src/main/java/com/google/samples/apps/nowinandroid/MainActivity.kt [hunk 6]
```diff
@@ -38,16 +58,60 @@ class MainActivity : ComponentActivity() {
...
 
         setContent {
-            NiaApp(calculateWindowSizeClass(this))
+            val systemUiController = rememberSystemUiController()
+            val darkTheme = shouldUseDarkTheme(userDataResult)
+
+            // Update the dark content of the system bars to match the theme
+            DisposableEffect(systemUiController, darkTheme) {
+                systemUiController.systemBarsDarkContentEnabled = !darkTheme
+                onDispose {}
```
Comment (tunjid) [inline app/src/main/java/com/google/samples/apps/nowinandroid/MainActivity.kt:104]: If we don't have anything to dispose, can we use a `LaunchedEffect`?

Comment (alexvanyo) [reply app/src/main/java/com/google/samples/apps/nowinandroid/MainActivity.kt:104]: Yes, although I think they are slightly different: From my understanding `DisposableEffect` wlil run slightly sooner than `LaunchedEffect` since `LaunchedEffect` requires a dispatch. I don't know if that is a large enough difference that this call will be delayed by a frame or not.

File: core/designsystem/src/main/java/com/google/samples/apps/nowinandroid/core/designsystem/theme/Theme.kt (hunk 3 above)
Comment (alexvanyo) [reply core/designsystem/src/main/java/com/google/samples/apps/nowinandroid/core/designsystem/theme/Theme.kt:194]: We also use it in the previews for `Background`

File: feature/bookmarks/src/main/java/com/google/samples/apps/nowinandroid/feature/bookmarks/BookmarksScreen.kt [hunk 7]
```diff
@@ -74,18 +71,6 @@ fun BookmarksScreen(
     modifier: Modifier = Modifier
 ) {
     Scaffold(
```
Comment (alexvanyo) [inline feature/bookmarks/src/main/java/com/google/samples/apps/nowinandroid/feature/bookmarks/BookmarksScreen.kt:73]: Nit: Once we remove the `topBar` here, it looks like the `Scaffold` doesn't do anything?

File: core/datastore/src/main/java/com/google/samples/apps/nowinandroid/core/datastore/NiaPreferencesDataSource.kt (hunk 2 above)
Comment (dturner) [reply core/datastore/src/main/java/com/google/samples/apps/nowinandroid/core/datastore/NiaPreferencesDataSource.kt]: Added a `null` branch to handle this.

File: core/data/src/test/java/com/google/samples/apps/nowinandroid/core/data/repository/OfflineFirstUserDataRepositoryTest.kt (hunk 1 above)
Comment (dturner) [reply core/data/src/test/java/com/google/samples/apps/nowinandroid/core/data/repository/OfflineFirstUserDataRepositoryTest.kt:152]: This is consistent the existing (weird) test naming so I've filed a bug here to fix this to avoid polluting this PR.  https://github.com/android/nowinandroid/issues/345

File: app/src/main/java/com/google/samples/apps/nowinandroid/MainActivity.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (dturner) [reply app/src/main/java/com/google/samples/apps/nowinandroid/MainActivity.kt]: Comment added.

File: feature/bookmarks/src/main/java/com/google/samples/apps/nowinandroid/feature/bookmarks/BookmarksScreen.kt (hunk 7 above)
Comment (dturner) [reply feature/bookmarks/src/main/java/com/google/samples/apps/nowinandroid/feature/bookmarks/BookmarksScreen.kt]: I removed it and couldn't see any changes to the layout so 👍.

File: app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt [hunk 8]
```diff
@@ -31,101 +31,160 @@ import androidx.compose.material3.ExperimentalMaterial3Api
...
+import com.google.samples.apps.nowinandroid.feature.settings.R as settingsR
+import com.google.samples.apps.nowinandroid.feature.settings.SettingsDialog
 import com.google.samples.apps.nowinandroid.navigation.NiaNavHost
 import com.google.samples.apps.nowinandroid.navigation.TopLevelDestination
 
 @OptIn(
     ExperimentalMaterial3Api::class,
     ExperimentalLayoutApi::class,
-    ExperimentalComposeUiApi::class
+    ExperimentalComposeUiApi::class, ExperimentalLifecycleComposeApi::class
```
Comment (manuelvicnt) [inline app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt:76]:
```suggestion
    ExperimentalComposeUiApi::class, 
    ExperimentalLifecycleComposeApi::class
```

File: app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt [hunk 9]
```diff
@@ -31,101 +31,160 @@ import androidx.compose.material3.ExperimentalMaterial3Api
...
+        windowSizeClass = windowSizeClass
+    ),
 ) {
-    NiaTheme {
-        val background: @Composable (@Composable () -> Unit) -> Unit =
-            when (appState.currentDestination?.route) {
-                ForYouDestination.route -> { content -> NiaGradientBackground(content = content) }
-                else -> { content -> NiaBackground(content = content) }
-            }
+    val background: @Composable (@Composable () -> Unit) -> Unit =
```
Comment (manuelvicnt) [inline app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt:87]: What's this complexity? Could this be a slot API instead? And the `content` inside it could be using `moveableContent` so that state isn't lost

File: app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt [hunk 10]
```diff
@@ -31,101 +31,160 @@ import androidx.compose.material3.ExperimentalMaterial3Api
...
+            modifier = Modifier.semantics {
+                testTagsAsResourceId = true
+            },
+            containerColor = Color.Transparent,
+            contentColor = MaterialTheme.colorScheme.onBackground,
+            contentWindowInsets = WindowInsets(0, 0, 0, 0),
+            snackbarHost = { SnackbarHost(snackbarHostState) },
+            topBar = {
+                // Show the top app bar on top level destinations.
+                val topLevelDestination =
```
Comment (manuelvicnt) [inline app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt:107]: Should this logic be moved to `appState`? Feels more appropriate

File: app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt [hunk 11]
```diff
@@ -31,101 +31,160 @@ import androidx.compose.material3.ExperimentalMaterial3Api
...
+                    NiaTopAppBar(
+                        titleRes = topLevelDestination.titleTextId,
+                        actionIcon = NiaIcons.Settings,
+                        actionIconContentDescription = stringResource(
+                            id = settingsR.string.top_app_bar_action_icon_description
+                        ),
+                        colors = TopAppBarDefaults.centerAlignedTopAppBarColors(
+                            containerColor = Color.Transparent
+                        ),
+                        onActionClick = { appState.toggleSettingsDialog(true) }
```
Comment (manuelvicnt) [inline app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt:119]: If you're toggling the value, you don't need to pass the new value in, do you? if not, call the method `setSettingsDialog(true)`?

File: app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt [hunk 12]
```diff
@@ -31,101 +31,160 @@ import androidx.compose.material3.ExperimentalMaterial3Api
...
+                    )
+                }
+            }
+        ) { padding ->
+
+            val isOffline by appState.isOffline.collectAsStateWithLifecycle()
+
+            // If user is not connected to the internet show a snack bar to inform them.
+            val notConnected = stringResource(R.string.for_you_not_connected)
+            LaunchedEffect(isOffline) {
```
Comment (manuelvicnt) [inline app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt:138]: This is going to be executed every time that `isOffline` changes or this composable is added to the Composition. Do you want to do that or just show it once?

File: app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt [hunk 13]
```diff
@@ -31,101 +31,160 @@ import androidx.compose.material3.ExperimentalMaterial3Api
...
+                //  content doesn't display behind it.
             }
         }
     }
 }
 
 @Composable
 private fun NiaNavRail(
-    destinations: List<TopLevelDestination>,
+    destinations: Collection<TopLevelDestination>,
```
Comment (manuelvicnt) [inline app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt:187]: Why this change?

File: app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaAppState.kt [hunk 14]
```diff
@@ -41,58 +46,84 @@ import com.google.samples.apps.nowinandroid.feature.foryou.navigation.ForYouDest
...
-    val windowSizeClass: WindowSizeClass
+    val coroutineScope: CoroutineScope,
+    val windowSizeClass: WindowSizeClass,
+    networkMonitor: NetworkMonitor,
 ) {
     val currentDestination: NavDestination?
         @Composable get() = navController
             .currentBackStackEntryAsState().value?.destination
 
+    private var _shouldShowSettingsDialog by mutableStateOf(false)
```
Comment (manuelvicnt) [inline app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaAppState.kt:78]: you can do 

```
var shouldShowSettingsDialog by mutableStateOf(false)
    private set
```

File: app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaAppState.kt [hunk 15]
```diff
@@ -41,58 +46,84 @@ import com.google.samples.apps.nowinandroid.feature.foryou.navigation.ForYouDest
...
+
     /**
-     * Top level destinations to be used in the BottomBar and NavRail
+     * Map of top level destinations to be used in the TopBar, BottomBar and NavRail. The key is the
+     * route.
      */
-    val topLevelDestinations: List<TopLevelDestination> = listOf(
-        TopLevelDestination(
+    val topLevelDestinations: Map<String, TopLevelDestination> = mapOf(
+        ForYouDestination.route to TopLevelDestination(
```
Comment (manuelvicnt) [inline app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaAppState.kt:102]: Why do you need the route? That'd force all routes to be public and exposed to the `app` module

File: core/designsystem/src/main/java/com/google/samples/apps/nowinandroid/core/designsystem/component/TopAppBar.kt [hunk 16]
```diff
@@ -32,7 +32,9 @@ import androidx.compose.runtime.Composable
 import androidx.compose.ui.Modifier
 import androidx.compose.ui.graphics.vector.ImageVector
 import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.semantics.contentDescription
 import androidx.compose.ui.tooling.preview.Preview
+import com.google.samples.apps.nowinandroid.core.designsystem.R
```
Comment (manuelvicnt) [inline core/designsystem/src/main/java/com/google/samples/apps/nowinandroid/core/designsystem/component/TopAppBar.kt:37]: The changes in this file can be reverted

File: core/designsystem/src/main/res/values/strings.xml [hunk 17]
```diff
@@ -18,4 +18,5 @@
     <string name="follow">Follow</string>
     <string name="unfollow">Unfollow</string>
     <string name="browse_topic">Browse topic</string>
+    <string name="screen_title">Screen title</string>
```
Comment (manuelvicnt) [inline core/designsystem/src/main/res/values/strings.xml:21]: Is this needed?

File: feature/settings/src/main/java/com/google/samples/apps/nowinandroid/feature/settings/SettingsDialog.kt [hunk 18]
```diff
@@ -0,0 +1,275 @@
...
+import com.google.samples.apps.nowinandroid.core.model.data.ThemeBrand.DEFAULT
+import com.google.samples.apps.nowinandroid.feature.settings.R.string
+import com.google.samples.apps.nowinandroid.feature.settings.SettingsUiState.Loading
+import com.google.samples.apps.nowinandroid.feature.settings.SettingsUiState.Success
+
+@ExperimentalLifecycleComposeApi
+@Composable
+fun SettingsDialog(
+    onDismiss: () -> Unit,
+    viewModel: SettingsViewModel = hiltViewModel()
```
Comment (manuelvicnt) [inline feature/settings/src/main/java/com/google/samples/apps/nowinandroid/feature/settings/SettingsDialog.kt:66]: Not sure if this should be a ViewModel. Can it be a plain class that has the same responsibilities? The problem with ViewModel here is that if you show multiple `SettingsDialog` under the same `ViewModelStoreOwner` / scope, they all will get the same instance of the ViewModel. That might not be ideal in some cases

File: app/src/androidTest/java/com/google/samples/apps/nowinandroid/ui/NiaAppStateTest.kt [hunk 19]
```diff
@@ -45,21 +55,43 @@ import org.junit.Test
 @OptIn(ExperimentalMaterial3WindowSizeClassApi::class)
 class NiaAppStateTest {
 
+    @get:Rule
+    val mainDispatcherRule = MainDispatcherRule()
```
[the remark this answers is not in the document]
Comment (dturner) [reply app/src/androidTest/java/com/google/samples/apps/nowinandroid/ui/NiaAppStateTest.kt:59]: Good spot. Thanks. Removed.

File: app/src/androidTest/java/com/google/samples/apps/nowinandroid/ui/NiaAppStateTest.kt [hunk 20]
```diff
@@ -76,23 +108,42 @@ class NiaAppStateTest {
     }
 
     @Test
-    fun niaAppState_destinations() {
+    fun niaAppState_destinations() = runTest {
```
[the remark this answers is not in the document]
Comment (dturner) [reply app/src/androidTest/java/com/google/samples/apps/nowinandroid/ui/NiaAppStateTest.kt:111]: It is needed because `NiaAppState` now accepts a coroutine scope for the `isOffline` state flow.

File: app/src/main/java/com/google/samples/apps/nowinandroid/MainActivity.kt [hunk 21]
```diff
@@ -38,16 +58,61 @@ class MainActivity : ComponentActivity() {
...
+    lateinit var networkMonitor: NetworkMonitor
+
+    val viewModel: MainActivityViewModel by viewModels()
+
     override fun onCreate(savedInstanceState: Bundle?) {
-        installSplashScreen()
+        val splashScreen = installSplashScreen()
         super.onCreate(savedInstanceState)
 
+        var uiState: MainActivityUiState by mutableStateOf(Loading)
```
[the remark this answers is not in the document]
Comment (dturner) [reply app/src/main/java/com/google/samples/apps/nowinandroid/MainActivity.kt:70]: Because `collectAsStateWithLifecycle` can only be called from a`Composable` and `onCreate` isn't one.

File: app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt (hunk 9 above)
Comment (dturner) [reply app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt:87]: This code isn't part of this PR so I've filed an issue to address it separately: https://github.com/android/nowinandroid/issues/349

File: app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaAppState.kt (hunk 14 above)
Comment (dturner) [reply app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaAppState.kt:78]: noice

File: app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaAppState.kt (hunk 15 above)
Comment (dturner) [reply app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaAppState.kt:102]: All routes are already public, they were just exposed as a `List` rather than `Map`. This is necessary so that the bottom bar can render all the destinations, and the top bar can obtain the screen title. 

The reason for changing to a `Map` is that it's O(1) to obtain the current top level destination, rather than O(N) to iterate through the list. 

Happy to discuss better ideas, it does feel a bit messy.

File: app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt (hunk 12 above)
Comment (dturner) [reply app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt:138]: This is the same behaviour as we had previously on the "ForYou" screen. 

"or this composable is added to the composition" <- this will only happen when `NiaApp` is recomposed, which iiuc is only during a configuration change, which is what we want, to re-show the snackbar. 

My overall opinion here is that the warning is too aggressive and should probably only be shown once, for a few seconds, but that's a job for another PR: https://github.com/android/nowinandroid/issues/350

File: app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt (hunk 11 above)
Comment (dturner) [reply app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt:119]: Good spot. There's a few other instances of this in our codebase. Bug filed: https://github.com/android/nowinandroid/issues/351

File: app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt (hunk 10 above)
Comment (dturner) [reply app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt:107]: Done.

File: core/designsystem/src/main/res/values/strings.xml (hunk 17 above)
Comment (dturner) [reply core/designsystem/src/main/res/values/strings.xml:21]: Removed

File: core/designsystem/src/main/java/com/google/samples/apps/nowinandroid/core/designsystem/component/TopAppBar.kt (hunk 16 above)
Comment (dturner) [reply core/designsystem/src/main/java/com/google/samples/apps/nowinandroid/core/designsystem/component/TopAppBar.kt:37]: Done.

File: app/src/androidTest/java/com/google/samples/apps/nowinandroid/ui/NiaAppStateTest.kt [hunk 22]
```diff
@@ -45,21 +55,43 @@ import org.junit.Test
...
+    }
+
+    @After
+    fun cleanup() {
+        testScope.cancel()
+    }
+
     @Test
-    fun niaAppState_currentDestination() {
+    fun niaAppState_currentDestination() = runTest {
```
[the remark this answers is not in the document]
Comment (dturner) [reply app/src/androidTest/java/com/google/samples/apps/nowinandroid/ui/NiaAppStateTest.kt:84]: See other comment, unfortunately it is required, even though we're doing nothing with those coroutines, the app state still initializes a StateFlow using a coroutine scope.

File: feature/settings/src/main/java/com/google/samples/apps/nowinandroid/feature/settings/SettingsDialog.kt (hunk 18 above)
Comment (dturner) [reply feature/settings/src/main/java/com/google/samples/apps/nowinandroid/feature/settings/SettingsDialog.kt:66]: https://github.com/android/nowinandroid/issues/352

File: app/src/main/java/com/google/samples/apps/nowinandroid/MainActivity.kt (hunk 6 above)
Comment (tunjid) [reply app/src/main/java/com/google/samples/apps/nowinandroid/MainActivity.kt:104]: It might not be, the dangling lambda threw me off for a bit. If the preference is for `DisposableEffect`, we should probs add a comment explaining why.

Comment (dturner) [reply app/src/main/java/com/google/samples/apps/nowinandroid/MainActivity.kt:104]: Leaving as-is for now.

File: app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaAppState.kt (hunk 15 above)
Comment (dturner) [reply app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaAppState.kt:104]: In the interest of getting this merged I'm going to leave as-is. Please lmk if you feel strongly about changing this (it might change with your nav change anyway).

File: app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt (hunk 12 above)
Comment (dturner) [reply app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt:138]: Please file an issue if you think this should be changed.

**Review by dturner — APPROVED** [review]

File: feature/foryou/src/main/java/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreen.kt [hunk 23]
```diff
@@ -135,119 +116,88 @@ internal fun ForYouScreen(
...
+        )
 
-            item(span = { GridItemSpan(maxLineSpan) }) {
+            /*item(span = { GridItemSpan(maxLineSpan) }) {
                 Column {
                     Spacer(modifier = Modifier.height(8.dp))
                     // Add space for the content to clear the "offline" snackbar.
+                    // TODO: Check that the Scaffold handles this correctly in NiaApp
                     if (isOffline) Spacer(modifier = Modifier.height(48.dp))
                     Spacer(Modifier.windowInsetsBottomHeight(WindowInsets.safeDrawing))
```
Comment (alexvanyo) [inline feature/foryou/src/main/java/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreen.kt:184]: Removing the `windowInsetsBottomHeight` here means that the cards might overlap with the navigation bar insets when the navigation rail is used. Can we bring that back (even if we leave the TODO for the spacer that accounts for the snackbar?)

File: feature/settings/src/main/java/com/google/samples/apps/nowinandroid/feature/settings/SettingsViewModel.kt [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+                    )
+                )
+            }
+            .stateIn(
+                scope = viewModelScope,
+                // Starting eagerly means the user data is ready when the SettingsDialog is laid out
+                // for the first time. Without this the layout is done using the "Loading" text,
+                // then replaced with the user editable fields once loaded, however, the layout
+                // height doesn't change meaning all the fields are squashed into a small
+                // scrollable column.
```
Comment (alexvanyo) [inline feature/settings/src/main/java/com/google/samples/apps/nowinandroid/feature/settings/SettingsViewModel.kt:54]: I think we're running into https://issuetracker.google.com/issues/221643630 here.

File: feature/foryou/src/main/java/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreen.kt (hunk 23 above)
Comment (dturner) [reply feature/foryou/src/main/java/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreen.kt]: Thanks for spotting this. My mistake. I've added everything except the spacer for the snackbar back in.

File: app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt [hunk 25]
```diff
@@ -31,93 +31,154 @@ import androidx.compose.material3.ExperimentalMaterial3Api
...
-    NiaTheme {
-        val background: @Composable (@Composable () -> Unit) -> Unit =
-            when (appState.currentDestination?.route) {
-                TopLevelDestination.FOR_YOU.name -> { content ->
-                    NiaGradientBackground(content = content)
-                }
-                else -> { content -> NiaBackground(content = content) }
+    val background: @Composable (@Composable () -> Unit) -> Unit =
+        when (appState.currentDestination?.route) {
+            TopLevelDestination.FOR_YOU.name -> {
```
Comment (alexvanyo) [inline app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt:89]: I don't think this changed with this PR, but this comparison will no longer ever be true. I think it needs to be comparing with `currentTopLevelDestination`:

```suggestion
        when (appState.currentTopLevelDestination) {
            TopLevelDestination.FOR_YOU -> {
```

File: app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt [hunk 26]
```diff
@@ -31,93 +31,154 @@ import androidx.compose.material3.ExperimentalMaterial3Api
...
+            },
+            containerColor = Color.Transparent,
+            contentColor = MaterialTheme.colorScheme.onBackground,
+            contentWindowInsets = WindowInsets(0, 0, 0, 0),
+            snackbarHost = { SnackbarHost(snackbarHostState) },
+            topBar = {
+                // Show the top app bar on top level destinations.
+                val destination = appState.currentTopLevelDestination
+                if (destination != null) {
+                    NiaTopAppBar(
```
Comment (alexvanyo) [inline app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt:112]: A side-effect of moving the top app bar to the outermost scope: The top app bar is now rendered over the rest of the content in the scaffold. This means that the top app bar is overlapping with the nav rail, and blocks tapping on the top-most navigation rail item.

I think `Modifier.zIndex` would be the most straightforward workaround?

File: feature/settings/src/main/java/com/google/samples/apps/nowinandroid/feature/settings/SettingsViewModel.kt (hunk 24 above)
Comment (dturner) [reply feature/settings/src/main/java/com/google/samples/apps/nowinandroid/feature/settings/SettingsViewModel.kt:54]: Good to know, thanks. I tried the workaround suggested in the bug which was to add 

```
properties = DialogProperties(usePlatformDefaultWidth = false)
```

But this resulted in the dialog filling all the available width and looked pretty strange. I've added a note to remove the current workaround when the bug is fixed.

File: app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt (hunk 26 above)
Comment (dturner) [reply app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt:113]: Constantly amazed at how you spot things like this. Added the suggested workaround. 

Also, we should really add a large screen device to our CI workflow. https://github.com/android/nowinandroid/issues/353

**Review by alexvanyo — APPROVED** [review]
LGTM!

File: app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt (hunk 26 above)
Comment (Cancan9190) [reply app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt:113]: Paketler:
WhatsApp: 2.23.8.76
Haritalar: 11.76.0300
CarStream: 2.0.5
YouTube Music: 5.54.52
Fermata Control: 1.8.11
Google Play Store: 35.4.10-21 [0] [PR] 524351809
Google Play hizmetleri: 23.15.56 (150400-527597497)
Fermata Auto: 1.9.1
Mesajlar: messages.android_20230329_00_RC03.phone_dynamic
Google: 14.16.27.28.arm64
USB Audio Player PRO: 6.1.0.4
fizy: 9.1.7
Signal: 6.18.4
Android Auto: 9.3.631434-release
Telegram: 9.6.3
Telegram: 9.6.2
Messenger: 406.0.0.13.115
Google Play Kitaplar: 2023.4.3.0.4
CarStream: 1.8.3

İzinler:
phone: açık
contacts: açık
locatıon: açık
sms: açık
mıcrophone: açık
calendar: açık
call_log: açık
nearby_devıces: açık
